kinode/packages/kns-indexer/kns-indexer/src/lib.rs

```diff
@@ -42,54 +44,228 @@ const DELAY_MS: u64 = 1_000; // 1s
     // last block we have an update from
     last_block: u64,
+    // kv handle
+    // includes keys and values for:
+    // "meta:chain_id", "meta:version", "meta:last_block", "meta:contract_address",
+    // "names:{namehash}" -> "{name}", "nodes:{name}" -> "{node_info}"
+    kv: Kv<String, Vec<u8>>, // todo: maybe serialize directly into known enum of possible types?
 }
 
-impl From<State> for WitState {
-    fn from(s: State) -> Self {
-        let contract_address: [u8; 20] = s.contract_address.into();
-        WitState {
-            chain_id: s.chain_id.clone(),
-            contract_address: contract_address.to_vec(),
-            names: s
-                .names
-                .iter()
-                .map(|(k, v)| (k.clone(), v.clone()))
-                .collect::<Vec<_>>(),
-            nodes: s
-                .nodes
-                .iter()
-                .map(|(k, v)| (k.clone(), v.clone().into()))
-                .collect::<Vec<_>>(),
-            last_block: s.last_block.clone(),
+impl State {
+    fn load(our: &Address) -> Self {
+        let kv: Kv<String, Vec<u8>> = match kv::open(our.package_id(), "kns_indexer", Some(10)) {
+            Ok(kv) => kv,
+            Err(e) => panic!("fatal: error opening kns_indexer key_value database: {e:?}"),
+        };
+
+        let mut state = Self {
+            kv,
+            version: 0,
+            last_block: 0,
+        };
+
+        // load or initialize chain_id
+        let chain_id = state.get_chain_id();
+        if chain_id == 0 {
+            state.set_chain_id(CHAIN_ID);
         }
+
+        // load or initialize contract_address
+        let contract_address = state.get_contract_address();
+        if contract_address
+            == eth::Address::from_str(KIMAP_ADDRESS)
+                .expect("Failed to parse KIMAP_ADDRESS constant")
+        {
+            state.set_contract_address(contract_address);
+        }
+
+        // load or initialize last_block
+        let last_block = state.get_last_block();
+        if last_block == 0 {
+            state.set_last_block(KIMAP_FIRST_BLOCK);
+        }
+
+        // load or initialize version
+        let version = state.get_version();
+        if version == 0 {
+            state.set_version(1); // Start at version 1
+        }
+
+        // update state struct with final values
+        state.version = state.get_version();
+        state.last_block = state.get_last_block();
+
+        println!(
+            "\n     🐦‍⬛  KNS Indexer State\n\
+             ▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔\n\
+                Version    {:>6}\n\
+                Last Block {:>6}\n\
+                Chain ID   {:>6}\n\
+                KIMAP      {}\n\
+             ▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁\n",
+            state.version,
+            state.last_block,
+            state.get_chain_id(),
+            state.get_contract_address().to_string()
+        );
+
+        state
+    }
+
+    fn meta_version_key() -> &'static str {
```

Comment:
Is it best practice to return a static str and then `.to_string()` it, as we do below?

Reply:
probably not, kv helper takes ownership to serialize, so made those helpers just return a String directly instead!